AST Optimal Target table for Card usage

otocephaly opened the pull request:
To whomever it may concern,

This is a work in progress; however, I appreciate all feedback given in relation to this.
Please read the description at the top of the relevant section for the related key assumptions before providing feedback because if after the notes it still isn't clear to you, then it won't be clear to the user either and I require that kind of feedback as well.
Please also note this will not be able to be merged in the future without merging PR #1649, as there is a fix in there in relation to DoT damage.

The below code outputs this table which is an accordion within the Arcana Logs table (currently placed at the bottom of the same section.
Collapsed view example:
![image](https://user-images.githubusercontent.com/89435408/158037395-d0461f3f-59f5-4233-ba52-bad91c19f589.png)

Expanded view example:
![image](https://user-images.githubusercontent.com/89435408/158037413-d4716795-45a1-42b5-a786-aaafe116868d.png)

Also, I updated the regular arcana logs table to include an x or checkmark depending on whether the technically optimal role was chosen

Example:
![image](https://user-images.githubusercontent.com/89435408/158037461-fc90c6df-d7a8-4a12-98ba-d57c8f7126a1.png)

Thank you for taking the time to read this and provide feedback in relation.

Cheers,
Otocephaly

otocephaly committed:
yarn extract

yarn extract

otocephaly committed:
[AST] need status

this update was necessary for the AST module optimal target

otocephaly committed:
[AST] Optimized Tracking DRAFT

base module for tracking optimized target. considered draft at this point until I am happy (which will be never)

otocephaly committed:
[AST] fix minor spelling

minor spelling fix

otocephaly committed:
[AST] DO NOT USE. wait until PR #1649

DO NOT USE. REVERT COMMIT instead. Use PR #1649

otocephaly committed:
[AST] optimal targets display

oop should only show 3

otocephaly commented:
tentative update (consolidating card played and time columns):
![image](https://user-images.githubusercontent.com/89435408/158039569-05e20a11-e2f7-42ab-92d9-92b8e85f7707.png)

suShirow commented:
Correct me if I'm wrong, but due to crit and RNG wouldn't damage numbers be meaningless? It's not like one can go back in time and pick the winning horse.
What would be useful would be potency (and calculated expected potency if it's a dot). 
The reason I suggest this is because this tool is meant to be 
a) a learning tool 
b) informational in a way that fflogs does not provide.

By showing potency of GCDs + Abilities captured, it would lean more toward a learning tool rather than just analysis for the sake of it.
I also absolutely want to avoid a scenario where it's used just to pick out the player with the best/worst gear/performance.

otocephaly committed:
Merge branch 'xivanalysis:endwalker' into arcana_timing

otocephaly committed:
Merge branch 'xivanalysis:endwalker' into arcana_timing

otocephaly committed:
yarn extract

yarn extract

otocephaly committed:
[AST] updates needed for crit DH

Updates to core needed for crit DH based on FFLogs API

otocephaly committed:
[AST] normalization based on crit DH

Adding in these changes to normalize crit / DH in damage figures used for AST card optimizing

otocephaly commented:
Hello!

Additional changes included adding in crit / DH figures to normalize events.
For the crit/DH calculation, it is first based on what fflogs can give us (which is only applicable for DoT events)
If there were no DoTs for the associated actor, then the crit / DH was calculated using an average over the fight and then removing the effective modifications due to other crit/DH adjusting events (such as chain strat).
For all damage events, there is an associated tagged crit/DH modifier which is used to help normalize. This is considered even in DoT cases since fflogs doesn't give the crit/DH at the time of application only the associated crit/DH from the actor.
There are additional considerations based on auto crit/auto DH abilities and statuses; however, for that list, refer to the top of the related section.

For the AST card normalization, the formula is as stated at the top. Crit/DH rates are taken effectively at the point the damage is dealt which is where the crit/DH modifications are utilized.
Assumptions for related calculations are shown typically at source; however, if it is not explained well enough, please let me know and I'll add more notes!

Here is an example of what it looks like now:
![image](https://user-images.githubusercontent.com/89435408/163906957-8f5e6dc9-f859-4512-9160-7509b64eb3ce.png)

Here is a console log of the related unmodified crit/DH rates by actor
![image](https://user-images.githubusercontent.com/89435408/163907025-0911f37c-6159-4dac-8cf2-e5534f248aec.png)

(actor list if needed)
![image](https://user-images.githubusercontent.com/89435408/163907063-ba587ba4-d708-4046-8ff2-0cbdddbb7382.png)

Same example without normalizing damage figures
![image](https://user-images.githubusercontent.com/89435408/163907136-5dfd3645-0d58-4e97-aa08-62f28bcd6653.png)

otocephaly committed:
[AST] oop missed some

missed some statuses that should have been copied down, but I should really just automate this instead of doing this commit

hintxiv commented on src/parser/jobs/ast/modules/ArcanaTracking/ArcanaDoTsGroup.ts:
```diff
@@ -0,0 +1,103 @@
+	'STORMBITE',
+
+	//MCH
+	'FLAMETHROWER',
```
Important to note here that ground DoTs like FT and Doton don't snapshot debuffs like Chain Strat - I think they need a slightly different code path.

otocephaly replied:
Updated! Thanks for this
I've moved all ground dots to a separate const with separate handling
See below commit

hintxiv commented on src/parser/jobs/ast/modules/ArcanaTracking/CritAndDHRates.tsx:
```diff
@@ -0,0 +1,422 @@
+	override initialise() {
+		//6.1 additions
+		if (this.parser.patch.after('6.08')) {
+			this.autoCritAbilities.push(this.data.actions.MIDARE_SETSUGEKKA.id)
+			this.autoCritAbilities.push(this.data.actions.KAESHI_SETSUGEKKA.id)
+			this.autoCritAbilities.push(this.data.actions.OGI_NAMIKIRI.id)
+			this.autoCritAbilities.push(this.data.actions.KAESHI_NAMIKIRI.id)
```
These should be in a const ideally.

otocephaly replied:
I've updated this, but let me know if it's the appropriate update
Thank you

hintxiv commented on src/parser/jobs/ast/modules/ArcanaTracking/CritAndDHRates.tsx:
```diff
@@ -0,0 +1,422 @@
+	]
+
+	//these are abilities that will automatically crit or direct hit which will skew results and should automatically be removed from the respective DH or Crit calculations
+	private autoCritAbilities: Array<Action['id']> = [
+		this.data.actions.STARFALL_DANCE.id,
+		this.data.actions.CHAOTIC_CYCLONE.id,
+		this.data.actions.INNER_CHAOS.id,
+		this.data.actions.PRIMAL_REND.id,
+	]
+	private autoDHAbilities: Array<Action['id']> = [
+		this.data.actions.STARFALL_DANCE.id,
+		this.data.actions.CHAOTIC_CYCLONE.id,
+		this.data.actions.INNER_CHAOS.id,
+		this.data.actions.PRIMAL_REND.id,
+	]
```
Wildfire needs some handling too. It can't crit or DH so the base damage is the only part that counts.

otocephaly replied:
Do you happen to know if wildfire is the only ability like this?

hintxiv replied:
I believe so yes

otocephaly replied:
Thanks! I added a section to handle this a bit generally in case this comes up again in the future
See below commit

hintxiv commented on src/parser/jobs/ast/modules/ArcanaTracking/CritAndDHRates.tsx:
```diff
@@ -0,0 +1,422 @@
+		[this.data.statuses.INNER_RELEASE.id]: {strength: 1},
+		//TO DO: these below moves only increase crit for the next weaponskill, not the oGCDs between
+		[this.data.statuses.REASSEMBLED.id]: {strength: 1},
+		[this.data.statuses.OPO_OPO_FORM.id]: {strength: 1},
```
Only BS and SotD count as auto crits when opo form is up, but it looks like this is going to count all damage (including other GCDs) when the form is up?

otocephaly replied:
Yeahhhh I was a bit lazy on this one!
I've updated this to have a separate section along with Inner Release for War as you've caught below
See below commit

hintxiv commented on src/parser/jobs/ast/modules/ArcanaTracking/CritAndDHRates.tsx:
```diff
@@ -0,0 +1,422 @@
+		[this.data.statuses.BATTLE_LITANY.id]: {strength: 0.1},
+		[this.data.statuses.DEVILMENT.id]: {strength: 0.2},
+		[this.data.statuses.THE_WANDERERS_MINUET.id]: {strength: 0.02},
+		[this.data.statuses.INNER_RELEASE.id]: {strength: 1},
```
Inner Release only applies to Fell Cleave and Decimate, but this will buff all damage including autos.

otocephaly replied:
Thanks for that catch!
I've updated this to have a separate section along with Opo for MNK
See below commit

hintxiv commented on src/parser/jobs/ast/modules/ArcanaTracking/CritAndDHRates.tsx:
```diff
@@ -0,0 +1,422 @@
+
+	private petDamage(event: Events['damage']) {
+		const owner: Actor['id'] | undefined = this.actors.get(event.source)?.owner?.id
+		//don't consider statuses as already considered in Damage and not expecting pets to have ticks
```
I think this is an unsafe assumption since pets have had DoTs in the past.

otocephaly replied:
Fair
Updated to include pet DoTs
See below commit

otocephaly committed:
[AST] update based on feedback

Based on Hint's feedback
- Updated to separate out ground dots which do not snapshot buffs/debuffs
- Further separated out crit modifiers based on cases which all have their own handling including crit/dh affected only on specific moves
- added status damage handling for pets
- Restructured init in crit/dh to fit a better format
- added crit types for easier handling in arcana optimized
- Moved Patch 6.1 to const
- Fixed crit rate for statuses which was bugging (order of operations changed since last commit?)

otocephaly commented:
Edit: there is a bug after reviewing the below. Will update when bug is resolved. Please hold off on reviews for now thx

Updated again!
(I should've put the last log link for retesting purposes, but alas here's another example structured similarly as before
https://www.fflogs.com/reports/a:PrCfLWbH7A2FQjx6)
![image](https://user-images.githubusercontent.com/89435408/164935458-9f831ff7-c736-4d35-b303-348407963377.png)


This version is after incorporating Hint's feedback as above
Current version:
![image](https://user-images.githubusercontent.com/89435408/164935585-4d167e67-9215-417e-9119-fd8177e1a479.png)

Console log of unmodified crits/dh by actor:
![image](https://user-images.githubusercontent.com/89435408/164935850-32f3e4ea-e2ca-4ae6-adab-f6a20b2df2a8.png)

Version without adjusting for crits/DH for comparative purposes:
![image](https://user-images.githubusercontent.com/89435408/164936238-9c5143ca-edca-4b71-bf78-a7d66c7d6d3b.png)

ayyaruq commented on src/event.ts:
```diff
@@ -161,6 +161,9 @@ interface EventDamage extends FieldsMultiTargeted {
 		/** Bonus multiplier on the damage from which positionals can be inferred */
 		bonusPercent: number
 	}>>
+	/** direct hit and crit rates provided by fflogs. Note: crit is provided as a stat and not the actual rate */
+	directHitRate?: number
+	expectedCritRate?: number
```
If crit is taken as a stat, shouldn't we be using the existing player stats stuff for it? And what about DH then?

ayyaruq commented on src/parser/jobs/ast/modules/ArcanaTracking/CritAndDHRates.tsx:
```diff
@@ -0,0 +1,475 @@
+
+const BASE_CRIT_PROBABILITY = 0.05 //5%
+const BASE_CRIT_MULTIPLIER = 0.4
+const ROUNDING_FACTOR = 0.001 //to match rounding used in game for crit rates
```
There shouldn't be any rounding, it's just integer flooring.

ayyaruq commented on src/parser/jobs/ast/modules/ArcanaTracking/ArcanaTrackingOptimized.tsx:
```diff
@@ -0,0 +1,482 @@
+
+const OPTIMAL_CARD_PERCENTAGE = 0.06
+const NOT_OPTIMAL_CARD_PERCENTAGE = 0.03
+const MAX_PLAYERS = 8 //won't be processed for alliance raids, hunt groups, etc.
```
Why not just get the proper party size?

ayyaruq reviewed (commented):
Idle comments while I was reading. I'll note that these modules are massive and needlessly complex tho, this'll need some heavy simplification before it gets merged. Most obvious thing for that is you're defining internal vars to use in hooks but they're not used for anything else, you can just keep the original const and `this.data.matchStatusId(statuses: StatusKey[])` type stuff and ignore the maps.

otocephaly committed:
Merge branch 'endwalker' into arcana_timing

otocephaly committed:
Merge branch 'endwalker' into arcana_timing

AzariahAeroborn commented:
Closing per needing to be redone on Dawntrail branch and updated for new cards.
